File: packages/backend-contract/src/domains/ephemeris.ts
/**
* Contract conventions:
* - Inputs are assumed validated at the backend boundary; the contract itself is primarily type-level.
* - Methods throw on invalid arguments or SPICE errors.
* - Lookups that may legitimately miss return `Found<T>` (`{ found: false }`) instead of throwing.
*/
import type {
  AbCorr,
  Found,
  SpiceHandle,
  SpiceStateVector,
  SpkezrResult,
  SpkposResult,
  VirtualOutput,
} from "../shared/types.js";
import type { SpiceIntCell, SpiceWindow } from "./cells-windows.js";

/**
 * Packed SPK segment descriptor (DAF summary) as a 5-double array.
 *
 * This is the packed form returned by `spksfs` and accepted by `spkuds`.
 */
export type SpkPackedDescriptor = readonly [number, number, number, number, number];

/** Unpacked SPK segment descriptor (see `spkuds_c`). */
export type SpkUnpackedDescriptor = {
  body: number;
  center: number;
  frame: number;
  type: number;

  first: number;
  last: number;

  baddr: number;
  eaddr: number;
};

export interface EphemerisApi {
  /**
   * Compute state relative to observer using loaded kernels (see `spkezr_c`).
   *
   * Note: `abcorr` is a known set of SPICE aberration correction strings, but we allow arbitrary
   * strings for forward-compatibility.
   */
  spkezr(
    target: string,
    et: number,
    ref: string,
    abcorr: AbCorr | string,
    observer: string,
  ): SpkezrResult;

  /**
   * Compute position relative to observer using loaded kernels (see `spkpos_c`).
   */
  spkpos(
    target: string,
    et: number,
    ref: string,
    abcorr: AbCorr | string,
    observer: string,
  ): SpkposResult;

  /** Compute state (6-vector) and light time via `spkez` (numeric IDs). */
  spkez(
    target: number,
    et: number,
    ref: string,
    abcorr: AbCorr | string,
    observer: number,
  ): SpkezrResult;

  /** Compute position (3-vector) and light time via `spkezp` (numeric IDs). */
  spkezp(
    target: number,
    et: number,
    ref: string,
    abcorr: AbCorr | string,
    observer: number,
  ): SpkposResult;

  /** Compute geometric state (6-vector) and light time via `spkgeo` (numeric IDs). */
  spkgeo(
    target: number,
    et: number,
    ref: string,
    observer: number,
  ): SpkezrResult;

  /** Compute geometric position (3-vector) and light time via `spkgps` (numeric IDs). */
  spkgps(
    target: number,
    et: number,
    ref: string,
    observer: number,
  ): SpkposResult;

  /** Compute state (6-vector) of `target` relative to the solar system barycenter via `spkssb`. */
  spkssb(target: number, et: number, ref: string): SpiceStateVector;

  /**
   * Compute the coverage window for an object in an SPK via `spkcov`.
   *
   * **Path semantics (backend-dependent):**
   * - Node backend: `spk` is a host filesystem path.
   * - WASM backend: `spk` is an Emscripten FS (virtual) path/id (typically the
   *   `path` you used in `furnsh({ path, bytes })`).
   *
   * **Output window requirements:** `cover` must be a valid, initialized window
   * handle created by the backend (e.g. `newWindow(maxIntervals)`) and must have
   * sufficient capacity for the merged output.
   *
   * **Error safety note:** Like CSPICE `spkcov_c`, if this routine throws while
   * updating `cover` (including due to insufficient capacity), the contents of
   * `cover` may be left in a corrupted/undefined state. If an error is thrown,
   * do **not** keep using `cover`; free it and create a fresh window.
   *
   * **Window semantics:** `cover` is updated in place. Like CSPICE `spkcov_c`,
   * coverage is **merged** with any intervals already present in `cover`.
   * Clear the window first (e.g. `scard(0, cover)`) if you want to avoid
   * accumulation.
   */
  spkcov(spk: string, idcode: number, cover: SpiceWindow): void;

  /**
   * Find the set of objects present in an SPK via `spkobj`.
   *
   * **Path semantics (backend-dependent):**
   * - Node backend: `spk` is a host filesystem path.
   * - WASM backend: `spk` is an Emscripten FS (virtual) path/id (typically the
   *   `path` you used in `furnsh({ path, bytes })`).
   *
   * **Output cell requirements:** `ids` must be a valid, initialized set cell
   * handle created by the backend (e.g. `newIntCell(size)`) and must have
   * sufficient capacity for the unioned output.
   *
   * **Error safety note:** Like CSPICE `spkobj_c`, if this routine throws while
   * updating `ids` (including due to insufficient capacity), the contents of
   * `ids` may be left in a corrupted/undefined state. If an error is thrown, do
   * **not** keep using `ids`; free it and create a fresh cell.
   *
   * **Cell semantics:** `ids` is updated in place. Like CSPICE `spkobj_c`, the
   * output is the **union** of the IDs already present in `ids` and the IDs
   * found in `spk`.
   * Clear the cell first (e.g. `scard(0, ids)`) if you want to
   * avoid accumulation.
   */
  spkobj(spk: string, ids: SpiceIntCell): void;

  /**
   * Search loaded SPK files for the highest-priority segment applicable to `body` and `et`.
   *
   * Note: `handle` is the native SPICE DAF handle for the file containing the segment.
   */
  spksfs(body: number, et: number): Found<{ handle: number; descr: SpkPackedDescriptor; ident: string }>;

  /** Pack an SPK segment descriptor via `spkpds`. */
  spkpds(
    body: number,
    center: number,
    frame: string,
    type: number,
    first: number,
    last: number,
  ): SpkPackedDescriptor;

  /** Unpack a packed SPK segment descriptor via `spkuds`. */
  spkuds(descr: SpkPackedDescriptor): SpkUnpackedDescriptor;

  // --- SPK writers ---------------------------------------------------------

  /**
   * Open a new SPK file for write (see `spkopn_c`).
   *
   * `file` interpretation is backend-dependent:
   * - Node: OS filesystem path
   * - WASM: virtual id under the backend's virtual filesystem (currently
   *   normalized into `/kernels/...`).
   *
   *   In other words, for the WASM backend, `file: string` is **not** a raw
   *   Emscripten absolute path. It is treated like other "kernel-ish" paths and
   *   is normalized into `/kernels`.
   *
   *   Examples (WASM backend):
   *   - `spkopn("out.bsp", ...)` writes to `/kernels/out.bsp`
   *   - `spkopn("/kernels/out.bsp", ...)` refers to the same file
   *   - `spkopn("/tmp/out.bsp", ...)` throws (OS paths/URLs are rejected)
   *
   * When `file` is a `VirtualOutput`, backends should allow reading bytes back
   * via `readVirtualOutput()` after closing the file handle.
   *
   * Callers should retain the `VirtualOutput` they passed to `spkopn`/`spkopa`.
   * It is the identifier used to read bytes back later.
   */
  spkopn(file: string | VirtualOutput, ifname: string, ncomch: number): SpiceHandle;

  /** Open an existing SPK for append (see `spkopa_c`). Same `file` semantics as `spkopn`. */
  spkopa(file: string | VirtualOutput): SpiceHandle;

  /** Close an SPK file previously opened by `spkopn`/`spkopa` (see `spkcls_c`). */
  spkcls(handle: SpiceHandle): void;

  /**
   * Write a type 8 SPK segment (see `spkw08_c`).
   *
   * `states` is a flat array with layout `[x,y,z, dx,dy,dz]` for each record.
   * The number of records `n` is derived as `states.length / 6`.
   */
  spkw08(
    handle: SpiceHandle,
    body: number,
    center: number,
    frame: string,
    first: number,
    last: number,
    segid: string,
    degree: number,
    states: readonly number[] | Float64Array,
    epoch1: number,
    step: number,
  ): void;
}

File: packages/backend-shim-c/src/domains/ephemeris.c
#include "tspice_backend_shim.h"
#include "tspice_error.h"

#include "SpiceUsr.h"

#include <string.h>

int tspice_spkezr(
    const char *target,
    double et,
    const char *ref,
    const char *abcorr,
    const char *observer,
    double *outState6,
    double *outLt,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }

  SpiceDouble state[6];
  SpiceDouble lt = 0.0;
  spkezr_c(target, (SpiceDouble)et, ref, abcorr, observer, state, &lt);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outState6) {
    outState6[0] = (double)state[0];
    outState6[1] = (double)state[1];
    outState6[2] = (double)state[2];
    outState6[3] = (double)state[3];
    outState6[4] = (double)state[4];
    outState6[5] = (double)state[5];
  }
  if (outLt) {
    *outLt = (double)lt;
  }

  return 0;
}

int tspice_spkpos(
    const char *target,
    double et,
    const char *ref,
    const char *abcorr,
    const char *observer,
    double *outPos3,
    double *outLt,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }

  SpiceDouble pos[3];
  SpiceDouble lt = 0.0;
  spkpos_c(target, (SpiceDouble)et, ref, abcorr, observer, pos, &lt);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outPos3) {
    outPos3[0] = (double)pos[0];
    outPos3[1] = (double)pos[1];
    outPos3[2] = (double)pos[2];
  }
  if (outLt) {
    *outLt = (double)lt;
  }

  return 0;
}


int tspice_spkopn(
    const char *path,
    const char *ifname,
    int ncomch,
    int *outHandle,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (err && errMaxBytes > 0) err[0] = '\0';
  if (outHandle) *outHandle = 0;

  if (!path || path[0] == '\0') {
    return tspice_return_error(err, errMaxBytes, "tspice_spkopn: path must be a non-empty string");
  }
  if (!ifname || ifname[0] == '\0') {
    return tspice_return_error(err, errMaxBytes, "tspice_spkopn: ifname must be a non-empty string");
  }
  if (ncomch < 0) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkopn: ncomch must be >= 0");
  }
  if (!outHandle) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkopn: outHandle must be non-NULL");
  }

  SpiceInt handleC = 0;
  spkopn_c(path, ifname, (SpiceInt)ncomch, &handleC);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  *outHandle = (int)handleC;
  return 0;
}

int tspice_spkopa(const char *path, int *outHandle, char *err, int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (err && errMaxBytes > 0) err[0] = '\0';
  if (outHandle) *outHandle = 0;

  if (!path || path[0] == '\0') {
    return tspice_return_error(err, errMaxBytes, "tspice_spkopa: path must be a non-empty string");
  }
  if (!outHandle) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkopa: outHandle must be non-NULL");
  }

  SpiceInt handleC = 0;
  spkopa_c(path, &handleC);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  *outHandle = (int)handleC;
  return 0;
}

int tspice_spkcls(int handle, char *err, int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (err && errMaxBytes > 0) err[0] = '\0';

  spkcls_c((SpiceInt)handle);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  return 0;
}

int tspice_spkw08(
    int handle,
    int body,
    int center,
    const char *frame,
    double first,
    double last,
    const char *segid,
    int degree,
    int n,
    const double *states6n,
    double epoch1,
    double step,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (err && errMaxBytes > 0) err[0] = '\0';

  if (!frame || frame[0] == '\0') {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08: frame must be a non-empty string");
  }
  if (!segid || segid[0] == '\0') {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08: segid must be a non-empty string");
  }
  if (n <= 0) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08: n must be > 0");
  }
  if (!states6n) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08: states6n must be non-NULL");
  }

  // Interpret `states6n` as an array of `n` 6-vectors.
  const SpiceDouble(*states)[6] = (const SpiceDouble(*)[6])states6n;

  spkw08_c(
      (SpiceInt)handle,
      (SpiceInt)body,
      (SpiceInt)center,
      frame,
      (SpiceDouble)first,
      (SpiceDouble)last,
      segid,
      (SpiceInt)degree,
      (SpiceInt)n,
      states,
      (SpiceDouble)epoch1,
      (SpiceDouble)step);

  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  return 0;
}

int tspice_spkw08_v2(
    int handle,
    int body,
    int center,
    const char *frame,
    double first,
    double last,
    const char *segid,
    int degree,
    int n,
    const double *states6n,
    int states6nLen,
    double epoch1,
    double step,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (err && errMaxBytes > 0) err[0] = '\0';

  if (!frame || frame[0] == '\0') {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08_v2: frame must be a non-empty string");
  }
  if (!segid || segid[0] == '\0') {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08_v2: segid must be a non-empty string");
  }
  if (n <= 0) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08_v2: n must be > 0");
  }
  if (!states6n) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08_v2: states6n must be non-NULL");
  }

  // Validate the flat input length before casting to `SpiceDouble[n][6]`.
  const int64_t expectedStates6nLen = (int64_t)n * 6;
  if ((int64_t)states6nLen != expectedStates6nLen) {
    return tspice_return_error(err, errMaxBytes, "tspice_spkw08_v2: states6nLen must equal 6*n");
  }

  // Interpret `states6n` as an array of `n` 6-vectors.
  const SpiceDouble(*states)[6] = (const SpiceDouble(*)[6])states6n;

  spkw08_c(
      (SpiceInt)handle,
      (SpiceInt)body,
      (SpiceInt)center,
      frame,
      (SpiceDouble)first,
      (SpiceDouble)last,
      segid,
      (SpiceInt)degree,
      (SpiceInt)n,
      states,
      (SpiceDouble)epoch1,
      (SpiceDouble)step);

  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  return 0;
}
